### PR TITLE
Add onlyChangedAttrs Test

### PR DESCRIPTION
**Pull request by mapedraza**
This PR adds the test described in #4330

### CODE REVIEW

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+onlyChangedAttrrs not working with csubs cache
+
+--SHELL-INIT--
+dbDrop ${CB_DB_NAME} tk
+brokerStart CB 0 IPv4 -subCacheIval 2 -multiservice
```
Comment (mapedraza) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:28]: Note this tests use sub cache

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+accumulatorStart --pretty-print localhost $LISTENER_PORT
+
+--SHELL--
+
+#
+# 01. Create entity E1 with attributes A and B
+# 02. Create subscription with onlyChangedAttrs true covering the entity
+# 03. Update E1-A
+# 04. Sleep until cache refres, update E1-A again
+# 05. Dump accumulator and see only A atribute in both notifications
```
Comment (mapedraza) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:38]: Steps 4 and 5 can be removed

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+                },
+                "type": "Text",
+                "value": "zzz"
+            },
+            "id": "E1",
+            "type": "T"
+        }
+    ],
+    "subscriptionId": "REGEX([0-9a-f]{24})"
+}
```
Comment (mapedraza) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:219]: The test expectations includes attribute `A` and `previousValue` metadata

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+     "value": 123,
+     "type": "Number"
+   }
+}'
+orionCurl --url /v2/entities --payload "$payload" --tenant tk --servicePath /
+echo
+echo
+
+
+echo '02. Create subscription with onlyChangedAttrs true covering the entity'
```
Comment (mapedraza) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:61]:
```suggestion
echo '02. Create subscription with onlyChangedAttrs true and special metadata'
```

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+============================================
+HTTP/1.1 201 Created
+Date: REGEX(.*)
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Location: /v2/entities/E1?type=T
+Content-Length: 0
+
+
+
+02. Create subscription with onlyChangedAttrs true covering the entity
```
Comment (mapedraza) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:131]:
```suggestion
02. Create subscription with onlyChangedAttrs true and special metadata
```

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test (hunk 2 above)
Comment (fgalan) [reply test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:38]: I think it doesn't hurt. I'd suggest to keep them. Cache is sometimes tricky, so better to have this covered.

NTC

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+# You should have received a copy of the GNU Affero General Public License
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+onlyChangedAttrrs not working with csubs cache
```
Comment (fgalan) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:24]:
```suggestion
Ensure onlyChangedAttrrs and previousMetadata work ok together
```

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+   "A": {
+     "value": "foo",
+     "type": "Text"
+   },
+   "B": {
+     "value": 123,
+     "type": "Number"
+   }
+}'
+orionCurl --url /v2/entities --payload "$payload" --tenant tk --servicePath /
```
Comment (fgalan) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:56]:
```suggestion
orionCurl --url /v2/entities --payload "$payload"
```

To keep test simpler, I'd suggest to remove tenant and servicepath. Also in `dbDrop` statements

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    "onlyChangedAttrs":true,
+    "metadata": [ "previousValue" ],
+    "throttling": 0
+  }
+}'
+orionCurl --url /v2/subscriptions --payload "$payload" --tenant tk --servicePath /
+echo
+echo
+
+sleep 2s
```
Comment (fgalan) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:84]: Why this sleep is needed? Have you observerd test glitching if not included?

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test (hunk 7 above)
Comment (fgalan) [reply test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:56]: `-multiservice` in brokerStart line can be also dropped

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test (hunk 8 above)
Comment (fgalan) [reply test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:84]: If removed, the sleep 2.5s statement in line 101 should be adjusted (to 4.5 I guess)

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+      }
+    ]
+   },
+  "notification": {
+    "http": {
+      "url": "http://localhost:'$LISTENER_PORT'/notify"
+    },
+    "onlyChangedAttrs":true,
+    "metadata": [ "previousValue" ],
+    "throttling": 0
```
Comment (fgalan) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:77]:
```suggestion
    "metadata": [ "previousValue" ]
```

To keep test simpler.

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test
Comment (fgalan) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:1]: I'd suggest to change file name to:

test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_and_previousvalue_built_in_metadata.test

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test (hunk 3 above)
Comment (fgalan) [reply test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:219]: Great! :)

NTC

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test (hunk 8 above)
Comment (mapedraza) [reply test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:84]: I almost copied an existing test. I did not change that value, so I am not aware about test glitching if not included. Should I remove it? This does not hurt

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test (hunk 7 above)
Comment (mapedraza) [reply test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:56]: Note: Removing tenant involves to change request headers expectations

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test
Comment (mapedraza) [reply test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test:1]: Done in https://github.com/telefonicaid/fiware-orion/pull/4335/commits/44d62c3b3c6a68ac7cd4aacff02bcf8ccb388c9c

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test (hunk 7 above)
Comment (mapedraza) [reply test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test]: Done in https://github.com/telefonicaid/fiware-orion/pull/4335/commits/44d62c3b3c6a68ac7cd4aacff02bcf8ccb388c9c

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test (hunk 8 above)
Comment (fgalan) [reply test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test]: For the sake of the test simplicity, I'd suggest to remove it (adjusting to 4.5s in line 101).

Comment (mapedraza) [reply test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_not_working_with_built_in_metadata.test]: Done in https://github.com/telefonicaid/fiware-orion/pull/4335/commits/4be9bdbad5d7c9722d6524d1cb19f843fbed6922

File: test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_and_previousvalue_built_in_metadata.test [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+Ensure onlyChangedAttrrs and previousMetadata work ok together
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB 0 IPv4 -subCacheIval 2
```
Comment (fgalan) [inline test/functionalTest/cases/4330_onlyChangedAttrs_built_in_metadata/onlychangedattrs_and_previousvalue_built_in_metadata.test:28]:
```suggestion
brokerStart CB 0 IPv4 -subCacheIval 4
```

**Review by fgalan — APPROVED** [review]
LGTM

**Pull request merged**